source/mesh_deformation/interface.cc

```diff
@@ -722,6 +740,253 @@ namespace aspect
+          if (!no_flux_boundary.empty() && sim.geometry_model->has_curved_elements())
+            {
+              AffineConstraints<double> user_level_constraints;
+              user_level_constraints.reinit(relevant_dofs);
+
+              internal::TangentialBoundaryFunctions::compute_no_normal_flux_constraints_shell(mesh_deformation_dof_handler,
+                                                                                              mg_constrained_dofs,
+                                                                                              mapping,
+                                                                                              level,
+                                                                                              0,
+                                                                                              no_flux_boundary,
+                                                                                              user_level_constraints);
+              user_level_constraints.close();
+              mg_constrained_dofs.add_user_constraints(level,user_level_constraints);
+
+              // let Dirichlet values win over no normal flux:
+              level_constraints.merge(user_level_constraints, AffineConstraints<double>::left_object_wins);
+              level_constraints.close();
+            }
+
+          typename MatrixFree<dim, double>::AdditionalData additional_data;
+          additional_data.tasks_parallel_scheme =
+            MatrixFree<dim, double>::AdditionalData::none;
+          additional_data.mapping_update_flags = update_flags;
+          additional_data.mg_level = level;
+          std::shared_ptr<MatrixFree<dim, double>> mg_mf_storage_level(
+                                                  new MatrixFree<dim, double>());
+
+          mg_mf_storage_level->reinit(mapping,
+                                      mesh_deformation_dof_handler,
+                                      level_constraints,
+                                      QGauss<1>(mesh_deformation_fe_degree + 1),
+                                      additional_data);
+          mg_matrices[level].clear();
+          mg_matrices[level].initialize(mg_mf_storage_level,
+                                        mg_constrained_dofs,
+                                        level);
+        }
+
+      MGTransferMatrixFree<dim, double> mg_transfer(mg_constrained_dofs);
+      mg_transfer.build(mesh_deformation_dof_handler);
+
+      using SmootherType =
+        PreconditionChebyshev<SystemMatrixType, dealii::LinearAlgebra::distributed::Vector<double>>;
+
+      mg::SmootherRelaxation<SmootherType, dealii::LinearAlgebra::distributed::Vector<double>> mg_smoother;
+
+      MGLevelObject<typename SmootherType::AdditionalData> smoother_data;
+      smoother_data.resize(0, n_levels - 1);
+
+      for (unsigned int level = 0; level < n_levels;
+           ++level)
+        {
+          if (level > 0)
```

Comment:
Please add some comments here on the choice of smoother_data. Why is it different depending on level (I guess coarse grid solver vs smoother on other levels, but we need this documented).

I guess you copied this from stokes_matrix_free.cc. Copy the comment describing the parameters as well.